### ISSUE TITLE
Optimize the has cache

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ipfs-blockstore#3]
The "has" cache uses a ton of memory. Go reports ~40MiB but that likely doesn't include allocator overhead. There are two things we may be able to do to help with this:

1. Only record that we *don't* have something. If we *do* have something, we'll have to load it anyways so caching the fact that we have it doesn't help much.
2. Use a better ARC cache. While the ARC cache was using ~40MiB, we only had ~7.5MiB worth of CIDs. Given (1), we may be able to make this more efficient with an ARC *set*.

Ideally, we'd use the bloom filter. Even for a 1% failure rate, at 1M items an optimal bloom filter would only be about a megabyte (requiring 7 lookups per check). However, loading the bloom filter is a bit of a pain.

**Comment by Kubuxu** [comment]
I had no idea that ARC cache uses so much memory, we should for sure reduce its size or switch to dumb LRU, RR or something else.

**Comment by lidel** [comment]
iiuc this was done at some point and code lives in  https://github.com/ipfs/boxo/blob/v0.10.1/blockstore/bloom_cache.go

### ❗ This repository is no longer maintained.

This repository is no longer maintained and this library has been copied over to [`boxo/blockstore`](https://github.com/ipfs/boxo/tree/main/blockstore).

In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
